lib/cache/cache.go

```diff
@@ -952,6 +980,111 @@ func (c *Cache) fetchAndWatch(ctx context.Context, retry utils.Retry, timer *tim
 	}
 }
 
+// performRelativeNodeExpiry performs a special kind of active expiration where we remove nodes
+// which are clearly stale relative to their more recently heartbeated counterparts as well as
+// the current time. This strategy lets us side-step issues of clock drift or general cache
+// staleness by only removing items which are stale from within the cache's own "frame of
+// reference".
+//
+// to better understand why we use this expiry strategy, its important to understand the two
+// distinct scenarios that we're trying to accommodate:
+//
+// 1. Expiry events are being emitted very lazily by the real backend (*hours* after the time
```

Comment:
We already filter expired items out of our dynamo queries.  The problem is with the event stream not keeping the cache up to date (since the expiry events can show up so late).

Active removal from the backend is a reasonable alternative, and a naive version of it was the first patch that was tried for the clusters that are currently experiencing issues due to this problem.  It resulted in very large spikes in backend usage and throttling errors.  While it is obviously possible to avoid this with the correct rate limiting, and by instructing users to up their backend limits, we're already seeing higher backend usage than we'd really like in large clusters.  Furthermore, concurrency is a factor in active removal from the backend, since auth servers would need to coordinate to limit concurrent deletion, and we'd need to have some protection against deletes racing with writes (not a super complex thing to fix, but it would require tinkering with our existing backend API as it doesn't really have a concept of conditional deletion at the moment).

Basically, I think this strategy is a safer bet for a time-sensitive patch release, since it doesn't affect shared state between auth servers, or run the risk of triggering backend throttling in clusters with tight margins.